src/wallet/db.cpp

```diff
@@ -72,11 +72,14 @@ BerkeleyEnvironment* GetWalletEnv(const fs::path& wallet_path, std::string& data
         database_filename = "wallet.dat";
     }
     LOCK(cs_db);
-    // Note: An ununsed temporary BerkeleyEnvironment object may be created inside the
-    // emplace function if the key already exists. This is a little inefficient,
-    // but not a big concern since the map will be changed in the future to hold
-    // pointers instead of objects, anyway.
-    return &g_dbenvs.emplace(std::piecewise_construct, std::forward_as_tuple(env_directory.string()), std::forward_as_tuple(env_directory)).first->second;
+    auto inserted = g_dbenvs.emplace(env_directory.string(), std::weak_ptr<BerkeleyEnvironment>());
+    if (inserted.second) {
+        auto env = std::make_shared<BerkeleyEnvironment>(env_directory.string());
+        inserted.first->second = env;
+        return env;
+    } else {
+        return inserted.first->second.lock();
```

Comment:
I may be missing the logic here, but how is it guaranteed that the `shared_ptr` hasn't expired yet?

Reply:
It's valid until `~BerkeleyEnvironment()` which means `shared_ptr` just expired, and in `~BerkeleyEnvironment()` the map is erased.

Reply:
Ah, of course, makes sense.